src/t8_schemes/t8_quads_w_rectangular_subelements/t8_subelements/t8_subelements_quad_cxx.cxx

```diff
@@ -0,0 +1,1523 @@
+   *      |             |         | 1 \   /   |        |   \   /   |        |   \   /   |        |   \ | /   |        |   \ | /   |
+   *      |             |   -->   x - - X   3 |   or   x - - x     |   or   x - - x - - x   or   x - - x - - x   or   x - - x - - x
+   *      |             |         | 0 /   \   |        |   / | \   |        |   /   \   |        |   /   \   |        |   / | \   |
+   *      | elem        |         | /   4   \ |        | /   |   \ |        | /       \ |        | /       \ |        | /   |   \ |
+   *      + - - - - - - x         x - - - - - x        x - - x - - x        x - - - - - x        x - - - - - x        x - - x - - x
+   *           
+   * Sub_ids are counted clockwise, starting with the (lower) left subelement with id 0.                    
+   * Note, that we do not change the p4est quadrant. */
+
+  int                 sub_id_counter = 0;
+  for (sub_id_counter = 0; sub_id_counter < num_subelements; sub_id_counter++) {
+    pquad_w_sub_subelement[sub_id_counter]->p4q.x = q->x;
+    pquad_w_sub_subelement[sub_id_counter]->p4q.y = q->y;
+    pquad_w_sub_subelement[sub_id_counter]->p4q.level = level;
+    pquad_w_sub_subelement[sub_id_counter]->dummy_is_subelement =
+      T8_IS_SUBELEMENT;
+    pquad_w_sub_subelement[sub_id_counter]->subelement_type = type;
+    pquad_w_sub_subelement[sub_id_counter]->subelement_id = sub_id_counter;
+  }
+
+  int                 i;
+  for (i = 0; i < num_subelements; ++i) {
+    T8_ASSERT (t8_element_is_valid (c[i]));
+    t8_element_copy_surround (q, &pquad_w_sub_subelement[i]->p4q);
+  }
+}
+
+int
+t8_default_scheme_sub_c::t8_element_get_number_of_subelements (int
+                                                               subelement_type,
+                                                               const
+                                                               t8_element *
+                                                               elem)
+{
+  T8_ASSERT (t8_element_is_valid (elem));
+
+  /* consider subelement_type 13 = 1101 in base two -> there are 4 + (1+1+0+1) = 7 subelements */
+  int                 num_subelements;
+  int                 type = subelement_type;
+  int                 digit_sum_of_binary = 0;
+  int                 binary_array[P4EST_FACES] = { };
+
+  int                 i;
+  int                 j;
+
+  for (i = 0; i < P4EST_FACES; i++) {   /* need an array with 4 elements to store all subelement types of the quad scheme from 1 to 15 ({0,0,0,1} to {1,1,1,1}) */
+    if (type >= pow (2, (P4EST_FACES - 1) - i)) {
+      binary_array[i] = 1;
+      type -= pow (2, (P4EST_FACES - 1) - i);
+    }
+  }                             /* we now got a binary represenation of the subelement type, bitwise stored in an array */
+
+  for (j = 0; j < P4EST_FACES; j++) {
+    digit_sum_of_binary += binary_array[j];
+  }
```

Comment:
If i understand this correctly you are counting the number of one bits of type?

So if type is 10, which is (1010) in binary, then digit_sum_of_binary = 2.
You can solve this way more efficient without using the binary_array.
Effectively type is already stored in binary and you should use that.
```  
  for (j = 0; j < P4EST_FACES; j++) {
    digit_sum_of_binary += !!(type & (1 << j));
  }
```
or, more readable, same logic
```  
  for (j = 0; j < P4EST_FACES; j++) {
    digit_sum_of_binary += (type & (1 << j)) ? 1 : 0;
  }
```
Explanation (example with type = 1010, j = 3):

- `1 << j` equal `2^j`   (2^3 = 8 = 1000 in binary)
- type & 2^j  make a bitwise AND of type and 2^j, essentaly this means the result is 0 if type has 0 in its j-th bit and the result is 2^j if it has a 1 in the j-th bit (1010 & 1000 = 1000)
- The last step is to convert this number to 1 if it is nonzero and 0 otherwise, either using two times NOT or using explicit computation (!!1000 = 1 or 1000 ? 1 : 0 yields 1)

Reply:
Thanks, I got it.